File: spring2013/codeWar/LRS/Clients/PlayerCSharpAI/AI/Utilities.cs
// Created by Windward Studios, Inc. (www.windward.net). No copyright claimed - do anything you want with this code.

using System;
using System.Collections.Generic;
using System.Drawing;
using PlayerCSharpAI.api;

namespace PlayerCSharpAI.AI
{
	/// <summary>
	/// Utilities that can be of use to an AI.
	/// </summary>
	public static class Utilities
	{

		/// <summary>
		/// Destination for a movement. Ignores all robots on the map but does take into account walls, conveyor belts and gears. Returns
		/// the final location of the move.
		/// </summary>
		/// <param name="map">The game map.</param>
		/// <param name="startLocation">Where the unit starts.</param>
		/// <param name="cards">The cards to apply.</param>
		/// <returns>The final location of the move.</returns>
		public static MovePoint CardDestination(GameMap map, BoardLocation startLocation, IEnumerable<Card> cards)
		{
			List<MovePoint> points = CardPath(map, startLocation, cards);
			if ((points == null) || (points.Count == 0))
				return null;
			return points[points.Count - 1];
		}

		/// <summary>
		/// Destination for a movement. Ignores all robots on the map but does take into account walls, conveyor belts and gears. Returns
		/// every step of the move.
		/// </summary>
		/// <param name="map">The game map.</param>
		/// <param name="startLocation">Where the unit starts.</param>
		/// <param name="cards">The cards to apply.</param>
		/// <returns>Every step of the move.</returns>
		public static List<MovePoint> CardPath(GameMap map, BoardLocation startLocation, IEnumerable<Card> cards)
		{
			// if we can't move, we end up where we started
			List<MovePoint> points = new List<MovePoint> {new MovePoint(startLocation)};

			foreach (Card cardOn in cards)
			{
				// move robot
				MovePoint endLocation = Move(map, startLocation, cardOn.Move);
				if (endLocation.Dead)
				{
					points.Add(endLocation);
					return points;
				}
				if (! endLocation.Location.Equals(startLocation))
				{
					startLocation = endLocation.Location;
					points.Add(new MovePoint(startLocation));
				}

				// conveyor belt - may cause a 1/4 turn.
				MapSquare sq = map.GetSquare(startLocation.MapPosition);
				for (int speed=1; (sq.Conveyor != null) && (speed <= sq.Conveyor.Speed); speed++)
				{
					endLocation = Move(map, startLocation.MapPosition, sq.Conveyor.Direction);
					BoardLocation locMove = new BoardLocation(endLocation.Location.MapPosition, startLocation.Direction);
					sq = map.GetSquare(endLocation.Location.MapPosition);
					if (sq.Conveyor != null)
					{
						MapSquare.DIRECTION dirEnter = MoveDirection(startLocation.MapPosition, endLocation.Location.MapPosition);
						locMove = locMove.Rotate((int)sq.Conveyor.Direction - (int)dirEnter);
					}
					startLocation = locMove;
					points.Add(new MovePoint(startLocation));
				}

				// gears
				if (sq.Type == MapSquare.TYPE.ROTATE_CLOCKWISE)
				{
					startLocation = startLocation.Rotate(1);
					points.Add(new MovePoint(startLocation));
				}
				if (sq.Type == MapSquare.TYPE.ROTATE_COUNTERCLOCKWISE)
				{
					startLocation = startLocation.Rotate(-1);
					points.Add(new MovePoint(startLocation));
				}

				// damage
				int damage = CalcLaserDamage(map, startLocation);
				if (damage != 0)
					points[points.Count - 1].Damage = damage;
			}

			return points;
		}

		/// <summary>
		/// Calculates what damage a unit will receive from lasers at a given location.
		/// </summary>
		/// <param name="map">The game map.</param>
		/// <param name="location">Where the unit is located.</param>
		/// <returns>The amount of damage. Will be 0 or 1.</returns>
		public static int CalcLaserDamage(GameMap map, BoardLocation location)
		{
			int damage = 0;
			damage += _CalcLaserDamage(map, location.MapPosition, 0, -1, MapSquare.DIRECTION.SOUTH, MapSquare.SIDE.NORTH, MapSquare.SIDE.SOUTH);
			damage += _CalcLaserDamage(map, location.MapPosition, 0, 1, MapSquare.DIRECTION.NORTH, MapSquare.SIDE.SOUTH, MapSquare.SIDE.NORTH);
			damage += _CalcLaserDamage(map, location.MapPosition, -1, 0, MapSquare.DIRECTION.EAST, MapSquare.SIDE.WEST, MapSquare.SIDE.EAST);
			damage += _CalcLaserDamage(map, location.MapPosition, 1, 0, MapSquare.DIRECTION.WEST, MapSquare.SIDE.EAST, MapSquare.SIDE.WEST);
			return damage;
		}

		private static int _CalcLaserDamage(GameMap map, Point position, int xAdd, int yAdd, MapSquare.DIRECTION laserDirection, MapSquare.SIDE wallExit, MapSquare.SIDE wallEnter)
		{
			int damage = 0;
			int x = position.X;
			int y = position.Y;
			bool startSquare = true;

			while ((0 <= x) && (x < map.Width) && (0 <= y) && (y < map.Height))
			{
				MapSquare sq = map.Squares[x][y];
				// can we move into this square?
				if ((! startSquare) && ((sq.Walls & wallEnter) != 0))
					break;
				startSquare = false;

				if ((sq.Laser != null) && (sq.Laser.Location.Direction == laserDirection))
				{
					damage++;
					break;
				}

				// can we move out of this square?
				if ((sq.Walls & wallExit) != 0)
					break;
				x += xAdd;
				y += yAdd;
			}
			return damage;
		}

		// NESW
		private static readonly MapSquare.SIDE[] sideMoveOut = { MapSquare.SIDE.NORTH, MapSquare.SIDE.EAST, MapSquare.SIDE.SOUTH, MapSquare.SIDE.WEST };
		private static readonly MapSquare.SIDE[] sideMoveIn = { MapSquare.SIDE.SOUTH, MapSquare.SIDE.WEST, MapSquare.SIDE.NORTH, MapSquare.SIDE.EAST };

		/// <summary>
		/// Move a unit one card move. Ignores all robots on the map but does take into account walls, conveyor belts and gears.
		/// </summary>
		/// <param name="map">The game map.</param>
		/// <param name="startLocation">Where the unit starts.</param>
		/// <param name="move">The move to apply.</param>
		/// <returns>The final location of the move.</returns>
		public static MovePoint Move(GameMap map, BoardLocation startLocation, Card.ROBOT_MOVE move)
		{
			int steps = 0;
			switch (move)
			{
				case Card.ROBOT_MOVE.BACKWARD_ONE:
					steps = -1;
					break;
				case Card.ROBOT_MOVE.FORWARD_ONE:
					steps = 1;
					break;
				case Card.ROBOT_MOVE.FORWARD_TWO:
					steps = 2;
					break;
				case Card.ROBOT_MOVE.FORWARD_THREE:
					steps = 3;
					break;
				case Card.ROBOT_MOVE.ROTATE_LEFT:
					return new MovePoint(startLocation.Rotate(-1));
				case Card.ROBOT_MOVE.ROTATE_RIGHT:
					return new MovePoint(startLocation.Rotate(1));
				case Card.ROBOT_MOVE.ROTATE_UTURN:
					return new MovePoint(startLocation.Rotate(2));
			}

			MapSquare.DIRECTION dir = steps >= 0 ? startLocation.Direction : startLocation.Rotate(2).Direction;
			Point position = startLocation.MapPosition;
			while (steps != 0)
			{
				MovePoint mp = Move(map, position, dir);
				if (mp.Dead)
					return new MovePoint(new BoardLocation(mp.Location.MapPosition, startLocation.Direction), true);
				position = mp.Location.MapPosition;
				int singleStep = Math.Max(-1, Math.Min(1, steps));
				steps -= singleStep;
			}
			return new MovePoint(new BoardLocation(position, startLocation.Direction));
		}

		/// <summary>
		/// Move a unit one square in the requested direction. Ignores all robots on the map but does take into account walls, conveyor belts and gears.
		/// </summary>
		/// <param name="map">The game map.</param>
		/// <param name="position">The map square to start the move from.</param>
		/// <param name="direction">The direction to move.</param>
		/// <returns>The final location of the move.</returns>
		public static MovePoint Move(GameMap map, Point position, MapSquare.DIRECTION direction)
		{

			// watch for wall in this direction
			MapSquare.SIDE sideExit = sideMoveOut[(int) direction];
			MapSquare.SIDE sideEnter = sideMoveIn[(int) direction];
			BoardLocation location = new BoardLocation(position, direction);

			// can we exit this square?
			MapSquare sq = map.GetSquare(position);
			if ((sq.Walls & sideExit) != 0)
				return new MovePoint(location);
			BoardLocation moveTo = location.Move(1);

			// did we go off the board?
			if ((moveTo.MapPosition.X < 0) || (map.Width <= moveTo.MapPosition.X) ||
				(moveTo.MapPosition.Y < 0) || (map.Height <= moveTo.MapPosition.Y))
				return new MovePoint(location, true);

			// did we go into a pit?
			if (map.GetSquare(moveTo.MapPosition).Type == MapSquare.TYPE.PIT)
				return new MovePoint(moveTo, true);

			// can we enter the new square?
			sq = map.GetSquare(moveTo.MapPosition);
			if ((sq.Walls & sideEnter) != 0)
				return new MovePoint(location);

			return new MovePoint(moveTo);
		}

		public static MapSquare.DIRECTION MoveDirection(Point start, Point end)
		{
			if (start.Y > end.Y)
				return MapSquare.DIRECTION.NORTH;
			if (start.Y < end.Y)
				return MapSquare.DIRECTION.SOUTH;
			if (start.X > end.X)
				return MapSquare.DIRECTION.WEST;
			if (start.X < end.X)
				return MapSquare.DIRECTION.EAST;
			throw new ApplicationException("start = end point " + start);
		}

		/// <summary>
		/// The result of one of the Utilities Move methods.
		/// </summary>
		public class MovePoint
		{
			/// <summary>
			/// Create the object. No damage and not dead.
			/// </summary>
			/// <param name="location">The location of the move result.</param>
			public MovePoint(BoardLocation location)
			{
				Location = location;
				Damage = 0;
				Dead = false;
			}

			/// <summary>
			/// Create the object. Not dead.
			/// </summary>
			/// <param name="location">The location of the move result.</param>
			/// <param name="damage">The damage level from this move.</param>
			public MovePoint(BoardLocation location, int damage)
			{
				Location = location;
				Damage = damage;
			}

			/// <summary>
			/// Create the object. No damage.
			/// </summary>
			/// <param name="location">The location of the move result.</param>
			/// <param name="dead">true if the move caused death.</param>
			public MovePoint(BoardLocation location, bool dead)
			{
				Location = location;
				Dead = dead;
			}

			/// <summary>
			/// The location of the move result.
			/// </summary>
			public BoardLocation Location { get; private set; }

			/// <summary>
			/// The damage level received from this move.
			/// </summary>
			public int Damage { get; set; }

			/// <summary>
			/// true if the move caused the unit to die (moved off the board).
			/// </summary>
			public bool Dead { get; private set; }
		}
	}
}

File: spring2013/codeWar/LRS/Clients/PlayerPythonAI/utilities.py
'''
Module utilities: contains utilities that can be of use to an AI.

MovePoint(location, damage, dead) -- the result of the following 'move' 
    functions.
cardDestination(gameMap, startLocation, cards) -- return the result of applying 
    cards in order, starting at startLocation.
cardPath(gameMap, startLocation, cards) -- like cardDestination, but returns a 
    list of every step along the way.
move(gameMap, location, movement) -- return the result of applying a single 
    movement at location.
moveDirection(start, end) -- calculate the direction of movement from start 
    to end.
calcLaserDamage(gameMap, location) -- calculate the damage a robot will 
    receive from lasers (not other robots!) at location.

Created on Dec 4, 2011

@author: malcolmm
'''

import api.player
from debug import trap
from api.board import BoardLocation, SIDE, TYPE
from api.board import DIRECTIONS as DIRS


class MovePoint(object):
    '''The result of one of the utilities module's "move" functions.'''
    
    def __init__(self, location, damage=0, dead=False):
        '''Create a MovePoint object where location is the result of the move.
        
        damage -- the amount of damage the move incurred on the robot. 
            Defaults to 0.
        dead -- whether the robot was killed (moved off the board).  Defaults 
            to False.
        
        '''
        self.location = location
        self.damage = damage
        self.dead = dead
    

def cardDestination(gameMap, startLocation, cards):
    '''Return the MovePoint resulting from applying cards at startLocation.
    
    Ignores all robots on the map but does take into account walls, conveyor 
    belts and gears.  Returns the final location of the move.
    
    gameMap -- the current state of the game map.
    startLocation -- the starting BoardLocation of the move.
    cards -- list of robot Card moves, to be applied in order.
    
    '''
    points = cardPath(gameMap, startLocation, cards)
    if points is None or len(points) == 0:
        trap()
        return None
    else:
        return points[-1]
    
def cardPath(gameMap, startLocation, cards):
    '''Return a list of MovePoints containing every step resulting from 
    applying cards at startLocation.
    
    Ignores all robots on the map but does take into account walls, 
    conveyor belts and gears.
    
    gameMap -- the current state of the game map.
    startLocation -- the starting BoardLocation of the move.
    cards -- list of robot Card moves, to be applied in order.
    
    '''
    #if we can't move, we end up where we started
    points = [MovePoint(startLocation)]
    
    for card in cards:
        #move robot
        endPoint = move(gameMap, startLocation, card.move)
        if endPoint.dead:
            points.append(endPoint)
            return points
        if endPoint.location != startLocation:
            startLocation = endPoint.location
            points.append(MovePoint(startLocation))
        
        sq = gameMap.getSquare(startLocation.mapPosition)
        
        #conveyor belt - may cause a 1/4 turn.
        speed = 1
        while sq.conveyor is not None and speed <= sq.conveyor.speed:
            endPoint = move(gameMap, startLocation.mapPosition, 
                            sq.conveyor.direction)
            locMove = BoardLocation(endPoint.location.mapPosition, 
                                    startLocation.direction)
            sq = gameMap.getSquare(endPoint.location.mapPosition)
            if sq.conveyor is not None:
                dirEnter = moveDirection(startLocation.mapPosition, 
                                         endPoint.location.mapPosition)
                if dirEnter is None:
                    dirEnter = startLocation.direction
                locMove = locMove.rotate(DIRS[sq.conveyor.direction] - DIRS[dirEnter])
            startLocation = locMove
            points.append(MovePoint(startLocation))
            speed += 1
        
        #gears
        if   sq.type == "ROTATE_CLOCKWISE":
            startLocation = startLocation.rotate(1)
        elif sq.type == "ROTATE_COUNTERCLOCKWISE":
            startLocation = startLocation.rotate(-1)
        else:
            assert sq.type in TYPE
        
        points.append(MovePoint(startLocation))
        
        #damage
        damage = calcLaserDamage(gameMap, startLocation)
        if damage != 0:
            points[-1].damage = damage
    return points

def calcLaserDamage(gameMap, location):
    '''Calculate the damage a unit will receive from lasers at location.'''
    damage = 0
    damage += _calcLaserDamage(gameMap, location.mapPosition, 0, -1, 
                               "SOUTH", SIDE["NORTH"], SIDE["SOUTH"])
    damage += _calcLaserDamage(gameMap, location.mapPosition, 0, 1, 
                               "NORTH", SIDE["SOUTH"], SIDE["NORTH"])
    damage += _calcLaserDamage(gameMap, location.mapPosition, -1, 0, 
                               "EAST", SIDE["WEST"], SIDE["EAST"])
    damage += _calcLaserDamage(gameMap, location.mapPosition, 1, 0, 
                               "WEST", SIDE["EAST"], SIDE["WEST"])
    return damage

def _calcLaserDamage(gameMap, position, xAdd, yAdd, laserDirection, wallExit, wallEnter):
    '''Return the damage a unit at position will take from lasers firing 
    from laserDirection.
    '''
    damage = 0
    x = position[0]
    y = position[1]
    startSquare = True
    
    while 0 <= x < gameMap.width and 0 <= y < gameMap.height:
        sq = gameMap.squares[x][y]
        #can we move into this square?
        if not startSquare and (sq.walls & wallEnter) != 0:
            break
        startSquare = False
        
        if sq.laser is not None and sq.laser.location.direction == laserDirection:
            damage += 1
            break #lasers cannot shoot through each other
        
        #can we move out of this square?
        if (sq.walls & wallExit) != 0:
            break
        
        x += xAdd
        y += yAdd
    return damage

def move(gameMap, location, movement):
    '''Return the MovePoint resulting from applying movement at location.
    
    Ignores all robots on the map but does take into account walls, conveyor 
    belts and gears.
    
    location -- Either a BoardLocation or (x, y) point.  If a BoardLocation, 
        'movement' will be considered a player.ROBOT_MOVE index.  If  a 
        tuple, 'movement' will be considered a board.DIRECTION index.
    movement -- A string indicating either the direction in which to move 
        (by one square), or the move card to apply.
    
    '''
    if   type(location) is type(tuple()):
        return _moveByDirection(gameMap, location, movement)
    elif location.__class__.__name__ == "BoardLocation":
        return _moveByCard(gameMap, location, movement)
    else:
        raise TypeError("Unexpected type for parameter 'location': %r" % location)

def _moveByCard(gameMap, startLocation, move):
    '''Move a unit one card move.

    Ignores all robots on the map but does take into account walls, 
    conveyor belts and gears.  Returns the final location of the move.
    '''
    steps = 0
    moves = api.player.ROBOT_MOVE
    assert len(moves) == 7
    if   move == "BACKWARD_ONE":
        steps = -1
    elif move == "FORWARD_ONE":
        steps = 1
    elif move == "FORWARD_TWO":
        steps = 2
    elif move == "FORWARD_THREE":
        steps = 3
    elif move == "ROTATE_LEFT":
        return MovePoint(startLocation.rotate(-1))
    elif move == "ROTATE_RIGHT":
        return MovePoint(startLocation.rotate(1))
    elif move == "ROTATE_UTURN":
        return MovePoint(startLocation.rotate(2))
    else:
        raise TypeError("Unrecognized move type %r" % move)
    
    moveDir = (startLocation.direction if steps >= 0 
               else startLocation.rotate(2).direction)
    position = startLocation.mapPosition
    while steps != 0:
        mp = _moveByDirection(gameMap, position, moveDir)
        if mp.dead:
            return MovePoint(BoardLocation(mp.location.mapPosition, 
                                           mp.location.direction), dead=True)
        position = mp.location.mapPosition
        singleStep = max(-1, min(1, steps))
        steps -= singleStep
    return MovePoint(BoardLocation(position, startLocation.direction))

#NESW
_sideMoveOut = ('NORTH', 'EAST', 'SOUTH', 'WEST')
_sideMoveIn  = ('SOUTH', 'WEST', 'NORTH', 'EAST')

def _moveByDirection(gameMap, position, direction):
    '''Move a unit one square in the given direction.
    
    Ignores all robots on the map but does take into account walls, conveyor 
    belts and gears.  Returns the final location of the move.
    '''
    #watch for wall in this direction
    sideExit  = SIDE[_sideMoveOut[DIRS[direction]]]
    sideEnter = SIDE[_sideMoveIn [DIRS[direction]]]
    location  = BoardLocation(position, direction)
    
    #can we exit this square?
    sq = gameMap.getSquare(position)
    if (sq.walls & sideExit) != 0:
        return MovePoint(location)
    moveTo = location.move(1)
    
    #did we go off the board?
    if (    moveTo.mapPosition[0] < 0 or gameMap.width  <= moveTo.mapPosition[0] or
            moveTo.mapPosition[1] < 0 or gameMap.height <= moveTo.mapPosition[1] or 
            gameMap.getSquare(moveTo.mapPosition).type == "PIT" ):
        return MovePoint(location, dead=True)
    
    #can we enter the new square?
    sq = gameMap.getSquare(moveTo.mapPosition)
    if (sq.walls & sideEnter) != 0:
        return MovePoint(location)
    
    return MovePoint(moveTo)

def moveDirection(start, end):
    '''Calculate the direction of movement from start to end.'''
    
    if   start[1] > end[1]:
        return "NORTH"
    elif start[1] < end[1]:
        return "SOUTH"
    elif start[0] > end[0]:
        return "WEST"
    elif start[0] < end[0]:
        return "EAST"
    else:
        raise TypeError("Start cannot equal end point (start: %r end: %r)" % 
                        (start, end))
        return None
    
